server/db/reindex.js

```diff
@@ -0,0 +1,154 @@
+    // edge case: if there was 1 result, postgres returns a singleton object, we want an array
+    if (!('length' in results)) {
+      results = [results];
+    }
+    log.debug('results.length is ' + results.length);
+    pgResultCount = results.length;
+    var esParams = [];
+    results.forEach(function(result) {
+      // bulk API is a little funny, two parts:
+      // 1. action description
+      esParams.push({ index: { _index: 'chronicle', _type: 'userPages', _id: result.id } });
+      // 2. document to index
+      esParams.push(result);
+    });
+    log.debug('esParams.length is ' + esParams.length);
+    return q(esParams);
+  };
+
+  var load_ = function(data) {
+    log.debug('inside load_, data.length is ' + data.length);
+    if (!data.length) {
+      // this is the edge case where no results came back - we're done
+      return q(0);
+    }
+    return elasticsearch.query('bulk', { body: data })
+      .then(function(esResponse) {
+        esResultCount = esResponse.items.length;
+        log.debug('esResultCount is ' + esResultCount + ', pgResultCount is ' + pgResultCount);
+        if (esResultCount === pgResultCount) {
+          log.debug('esResultCount is ' + esResultCount);
+          return q(esResultCount);
+        } else {
+          // something went wrong, try again
+          throw new Error('result counts did not match');
+        }
+      }, function(err) {
+        log.warn(err);
+        throw new Error('elasticsearch insertion failed');
+      });
+  };
+
+  q().timeout(config.get('db_reindex_timeout'))
+    .then(extract_)
+    .then(transform_)
+    .then(load_)
+    .done(function(result) {
+      log.debug('completed an iteration');
+      log.debug('result count from last iteration is ' + result);
+      // if the number of results was smaller than the stepSize, then that
+      // was the last step; we're done.
+      if (result < params.stepSize) {
+        isRunning = false;
+        log.debug('last iteration processed ' + result + ' records, step size is ' + params.stepSize);
+        log.info('reindexing complete');
+        params.done();
+        return process.exit();
+      }
+      // else, update the offset counter, then call this function again.
+      params.offset += params.stepSize;
+      params.errorCount = 0;
+      log.debug('reindexing another chunk, updated offset is now ' + params.offset);
+      job(params);
+    }, function(err) {
+      log.warn(err);
+      params.errorCount += 1;
+      if (params.errorCount > params.maxRetries) {
+        isRunning = false;
+        var msg = 'reindexing failed too many times, aborting';
+        log.critical(msg);
+        return params.done(msg);
+      }
+      setTimeout(function() { job(params); }, params.errorCount * params.retryWait);
+    });
+};
+
+function start(cb) {
+  log.info('starting elasticsearch reindex process');
+  if (isRunning) {
+    var msg = 'reindex is already running';
+    log.critical(msg);
+    return cb(msg);
```

Comment:
do we want to call the callback with a string error message, or should we wrap these in an Error constructor type thing?

```js
var err = new Error('reindex is already running');
log.critical(err.message);
return cb(err);
...
```